### ISSUE TITLE
Add spectral density computation using lanczos vectors

### ISSUE THREAD

**Issue by noahgolmant** [noahgolmant/pytorch-hessian-eigenthings#31]
Implement Stochastic Lanczos Quadrature using scipy or a custom Lanczos implementation. Integrates the approach from [this](https://github.com/google/spectral-density) repo.

**Comment by devansh20la** [comment]
Hi @noahgolmant,

I would like to work on this issue. Do you have some starting points?

Thanks

**Comment by noahgolmant** [comment]
Hello! Algorithm 1 from [this paper](https://arxiv.org/abs/1901.10159) has a mathematical description of the approach, and the authors then implemented it in the repo I linked above. I think the jax implementation is nice and clear. Most of the jax code should be one-to-one with PyTorch. The main steps would be:

1. Port the [lanczos](https://github.com/google/spectral-density/blob/master/jax/lanczos.py#L27) implementation to PyTorch. I am currently using scipy's built-in Lanczos method, but this only gives access to the final eigenvalues and eigenvectors, not the tridiagonal matrices.
2. Port the code for [density estimation from tridiagonal matrices](https://github.com/google/spectral-density/blob/master/jax/density.py#L120). 
3. Hook up the code to use the existing `HVPOperator` class with this to go straight from model/dataloader/loss to eigenvalue density.
4. Add some basic precision tests with random matrices.

Would love to see this happen!
